src/rcc/rec.rs

```diff
@@ -0,0 +1,256 @@
+//! Peripheral Reset and Enable Control (REC)
+//!
+//! This module contains safe accessors to the RCC functionality for each
+//! periperal.
+//!
+//! At a minimum each peripheral implements
+//! [ResetEnable](trait.ResetEnable.html). Additionally those peripherals
+//! that have a multiplexer in the PKSU may also have methods
+//! `kernel_clk_mux` and `get_kernel_clk_mux`. These set and get the state
+//! of the kernel clock multiplexer respectively.
+//!
+//! # Example
+//!
+//! ```
+//! // Constrain and Freeze power
+//! ...
+//! let rcc = dp.RCC.constrain();
+//! let ccdr = rcc.sys_ck(100.mhz()).freeze(vos, &dp.SYSCFG);
+//!
+//! // Enable the clock to a peripheral and reset it
+//! ccdr.peripheral.FDCAN.enable().reset();
+//! ```
+#![deny(missing_docs)]
+
+use core::marker::PhantomData;
+
+use super::Rcc;
+use crate::stm32::{rcc, RCC};
+use cortex_m::interrupt;
+
+/// A trait for Resetting, Enabling and Disabling a single peripheral
+pub trait ResetEnable {
+    /// Enable this peripheral
+    fn enable(self) -> Self;
+    /// Disable this peripheral
+    fn disable(self) -> Self;
+    /// Reset this peripheral
+    fn reset(self) -> Self;
+}
+
+impl Rcc {
+    /// Returns all the peripherals resets / enables / kernel clocks.
+    ///
+    /// # Use case
+    ///
+    /// Allows peripherals to be reset / enabled before the calling
+    /// freeze. For example, the internal watchdog could be enabled to
+    /// issue a reset if the call the freeze hangs waiting for an external
+    /// clock that is stopped.
+    ///
+    /// # Safety
+    ///
+    /// If this method is called multiple times, or is called before the
+    /// [freeze](struct.Rcc.html#freeze), then multiple accesses to the
+    /// same memory exist.
+    #[inline]
+    pub unsafe fn steal_peripheral_rec(&self) -> PeripheralREC {
+        PeripheralREC::new_singleton()
+    }
+}
+
+macro_rules! peripheral_reset_and_enable_control {
+    ($($AXBn:ident, $axb_doc:expr => [
+        $(
+            $( #[ $pmeta:meta ] )*
+                $p:ident
+            $( kernel$(($Variant:ident))* $ccip:ident $clk_doc:expr )*
+        ),*
+    ];)+) => {
+        paste::item! {
+            /// Peripheral Reset and Enable Control
+            #[allow(non_snake_case)]
+            pub struct PeripheralREC {
+                $(
+                    $(
+                        #[allow(missing_docs)]
+                        $( #[ $pmeta ] )*
+                        pub [< $p:upper >]: $p,
+                    )*
+                )+
+
+                // Private field making `Peripherals` non-exhaustive. We
+                // don't use `#[non_exhaustive]` so we can support older
+                // Rust versions.
```

Comment:
Maybe add a TODO here so we can fix this once `#[non_exhaustive]`has been around for a while.

Reply:
Oh and we should specify how long "a while" should take. So we dont just forget about it.

Reply:
We only need to bump the MSRV by one..